modules/client/src/internal/utils.ts

```diff
@@ -824,3 +863,801 @@ export function toPluginPermissionOperationType(
+        .INVALID_GRANT_ROOT_PERMISSION_PERMISSION,
+    );
+  }
+  if (
+    decodedPermission.permissionId !==
+      PermissionIds.ROOT_PERMISSION_ID
+  ) {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_GRANT_ROOT_PERMISSION_PERMISSION_ID,
+    );
+  }
+  return causes;
+}
+export function validateRevokeRootPermissionAction(
+  action: DaoAction,
+  daoAddress: string,
+  pspAddress: string,
+): PluginUpdateProposalInValidityCause[] {
+  const causes: PluginUpdateProposalInValidityCause[] = [];
+  const decodedPermission = decodeRevokeAction(action.data);
+  if (action.value.toString() !== "0") {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_REVOKE_ROOT_PERMISSION_VALUE,
+    );
+  }
+  if (decodedPermission.where !== daoAddress) {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_REVOKE_ROOT_PERMISSION_WHERE_ADDRESS,
+    );
+  }
+  if (decodedPermission.who !== pspAddress) {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_REVOKE_ROOT_PERMISSION_WHO_ADDRESS,
+    );
+  }
+  if (
+    decodedPermission.permission !== Permissions.ROOT_PERMISSION
+  ) {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_REVOKE_ROOT_PERMISSION_PERMISSION,
+    );
+  }
+  if (
+    decodedPermission.permissionId !==
+      PermissionIds.ROOT_PERMISSION_ID
+  ) {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_REVOKE_ROOT_PERMISSION_PERMISSION_ID,
+    );
+  }
+  return causes;
+}
+/**
+ * Validate a plugin update proposal
+ *
+ * @export
+ * @param {DaoAction} action
+ * @param {string} daoAddress
+ * @param {IClientGraphQLCore} graphql
+ * @param {IClientIpfsCore} ipfs
+ * @return {*}  {Promise<PluginUpdateProposalInValidityCause[]>}
+ */
+export async function validateApplyUpdateFunction(
+  action: DaoAction,
+  daoAddress: string,
+  graphql: IClientGraphQLCore,
+  ipfs: IClientIpfsCore,
+): Promise<PluginUpdateProposalInValidityCause[]> {
+  const causes: PluginUpdateProposalInValidityCause[] = [];
+  if (action.value.toString() !== "0") {
+    causes.push(
+      PluginUpdateProposalInValidityCause.INVALID_APPLY_UPDATE_ACTION_VALUE,
+    );
+  }
+  const decodedParams = decodeApplyUpdateAction(
+    action.data,
+  );
+  // get dao with plugins
+  type U = { dao: SubgraphDao };
+  const { dao } = await graphql.request<U>({
+    query: QueryDao,
+    params: { address: daoAddress },
+    name: "dao",
+  });
+  // find the plugin with the same address
+  const plugin = dao.plugins.find((plugin) =>
+    plugin.appliedPreparation?.pluginAddress ===
+      decodedParams.pluginAddress
+  );
+  if (plugin) {
+    // check release is the same as the one installed
+    if (
+      plugin.appliedVersion?.release.release !==
+        decodedParams.versionTag.release
+    ) {
+      causes.push(PluginUpdateProposalInValidityCause.INVALID_PLUGIN_RELEASE);
```

Comment:
INVALID_PLUGIN_RELEASE is not a good name.
Maybe: UPDATE_TO_INCOMPATIBLE_RELEASE